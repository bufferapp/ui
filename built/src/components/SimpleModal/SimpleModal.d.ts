import React from 'react';
import PropTypes from 'prop-types';
declare const SimpleModal: {
    ({ children, closeAction }: {
        children: any;
        closeAction: any;
    }): React.JSX.Element;
    propTypes: {
        children: PropTypes.Validator<string | number | boolean | PropTypes.ReactElementLike | PropTypes.ReactNodeArray>;
        closeAction: PropTypes.Validator<(...args: any[]) => any>;
    };
};
export default SimpleModal;
//# sourceMappingURL=SimpleModal.d.ts.map