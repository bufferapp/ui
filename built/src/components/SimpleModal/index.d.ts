export { default } from './SimpleModal';
//# sourceMappingURL=index.d.ts.map