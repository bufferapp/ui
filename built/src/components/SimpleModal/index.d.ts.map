{"version":3,"file":"index.d.ts","sourceRoot":"","sources":["../../../../src/components/SimpleModal/index.ts"],"names":[],"mappings":"AAAA,OAAO,EAAE,OAAO,EAAE,MAAM,eAAe,CAAA"}