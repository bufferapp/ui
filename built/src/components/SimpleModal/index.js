export { default } from './SimpleModal';
//# sourceMappingURL=index.js.map