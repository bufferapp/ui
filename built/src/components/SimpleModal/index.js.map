{"version":3,"file":"index.js","sourceRoot":"","sources":["../../../../src/components/SimpleModal/index.ts"],"names":[],"mappings":"AAAA,OAAO,EAAE,OAAO,EAAE,MAAM,eAAe,CAAA","sourcesContent":["export { default } from './SimpleModal'\n"]}