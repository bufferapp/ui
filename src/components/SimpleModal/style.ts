import styled, { keyframes } from 'styled-components'
import { easeOutQuart } from '../style/animations'
import { white, red } from '../style/colors'

const fadeIn = keyframes`
  0% {
    opacity: 0;
  }

  100% {
    opacity: 1;
  }
`

export const Container = styled.div<{ NonDismissibleModal: boolean }>`
  background: rgba(0, 0, 0, 0.7);
  display: flex;
  height: 100%;
  align-items: center;
  justify-content: center;
  position: fixed;
  box-sizing: border-box;
  width: 100%;
  z-index: ${({ NonDismissibleModal }): string =>
    NonDismissibleModal ? '1' : '9999'};
  flex-direction: column;
  top: 0px;
  left: 0px;

  animation: 200ms ${fadeIn} ${easeOutQuart};
`

export const Modal = styled.div`
  background-color: ${white};
  border-radius: 8px;
  box-shadow: 0px 1px 4px rgba(0, 0, 0, 0.16);
  box-sizing: border-box;
  margin: 0 0 1rem;
  width: fit-content;
  overflow: hidden;
  position: absolute;

  display: flex;
  align-items: center;
  justify-content: center;
  outline: none;
`

export const CloseButton = styled.button`
  background: none;
  border: none;
  position: absolute;
  top: 16px;
  right: 16px;
  cursor: pointer;
  padding: 0;
  z-index: 1;

  transition: transform 0.15s ease-out;
  svg {
    transition: fill 0.15s ease-out;
  }

  &:hover {
    transform: scale(1.1);
    svg {
      fill: ${red};
    }
  }
`
