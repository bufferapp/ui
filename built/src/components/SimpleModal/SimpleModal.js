import React, { useRef, useEffect } from 'react';
import PropTypes from 'prop-types';
import { Cross } from '../Icon';
import { Container, Modal, CloseButton } from './style';
import { useAnimation } from '../AnimationWrapper';
import { stageInCenter, stageOutCenter } from '../style/animations';
const ESCAPE_KEY = 27;
const TAB_KEY = 9;
const SimpleModal = ({ children, closeAction }) => {
    const modalRef = useRef(null);
    const containerRef = useRef(null);
    const { AnimationWrapper, dismiss: dismissAnimationWrapper, animationProps, } = useAnimation({
        stageInAnimation: stageInCenter,
        stageOutAnimation: stageOutCenter,
        duration: 400,
        onDismiss: closeAction,
    });
    const handleTabKey = (e) => {
        const focusableModalElements = modalRef.current.querySelectorAll('a[href], button, textarea, input[type="text"], input[type="radio"], input[type="checkbox"], select, div[tabIndex="0"]');
        const firstElement = focusableModalElements[0];
        const lastElement = focusableModalElements[focusableModalElements.length - 1];
        if (!e.shiftKey && document.activeElement === lastElement) {
            firstElement.focus();
            return e.preventDefault();
        }
        if (e.shiftKey && document.activeElement === firstElement) {
            lastElement.focus();
            e.preventDefault();
        }
    };
    const keyListenersMap = new Map([
        [ESCAPE_KEY, dismissAnimationWrapper],
        [TAB_KEY, handleTabKey],
    ]);
    const clickToClose = (e) => {
        if (e.target !== containerRef.current)
            return;
        dismissAnimationWrapper();
    };
    useEffect(() => {
        function keyListener(e) {
            const listener = keyListenersMap.get(e.keyCode);
            return listener && listener(e);
        }
        document.addEventListener('keydown', keyListener);
        modalRef.current.focus();
        containerRef.current.addEventListener('click', (e) => clickToClose(e));
        return () => document.removeEventListener('keydown', keyListener);
    }, []);
    return (React.createElement(Container, { ref: containerRef, role: "dialog", "aria-modal": "true" },
        React.createElement(AnimationWrapper, Object.assign({}, animationProps),
            React.createElement(Modal, { ref: modalRef, tabIndex: "0" },
                React.createElement(CloseButton, { onClick: () => dismissAnimationWrapper() },
                    React.createElement(Cross, null)),
                children))));
};
SimpleModal.propTypes = {
    children: PropTypes.node.isRequired,
    closeAction: PropTypes.func.isRequired,
};
export default SimpleModal;
//# sourceMappingURL=SimpleModal.js.map