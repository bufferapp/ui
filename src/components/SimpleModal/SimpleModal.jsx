import React, { useRef, useEffect } from 'react';
import PropTypes from 'prop-types';
import  Cross  from '../Icons/Cross';
import { Container, Modal, CloseButton } from './style';
import { useAnimation } from '../AnimationWrapper';
import { stageInCenter, stageOutCenter } from '../style/animations';

const ESCAPE_KEY = 27;
const TAB_KEY = 9;
const SimpleModal = ({ children, closeAction }) => {
  const modalRef = useRef(null);
  const containerRef = useRef(null);
  const { AnimationWrapper, dismiss:dismissAnimationWrapper, animationProps } = useAnimation({
    stageInAnimation: stageInCenter,
    stageOutAnimation: stageOutCenter,
    duration: 400,
    onDismiss: closeAction,
  })

  const handleTabKey = e => {
    const focusableModalElements = modalRef.current.querySelectorAll(
      'a[href], button, textarea, input[type="text"], input[type="radio"], input[type="checkbox"], select, div[tabIndex="0"]'
    );
    const firstElement = focusableModalElements[0];
    const lastElement =
      focusableModalElements[focusableModalElements.length - 1];

    if (!e.shiftKey && document.activeElement === lastElement) {
      firstElement.focus();
      return e.preventDefault();
    }

    if (e.shiftKey && document.activeElement === firstElement) {
      lastElement.focus();
      e.preventDefault();
    }
  };

  const keyListenersMap = new Map([
    [ESCAPE_KEY, dismissAnimationWrapper],
    [TAB_KEY, handleTabKey],
  ]);

  const clickToClose = e => {
    if (e.target !== containerRef.current) return;
    dismissAnimationWrapper();
  };

  useEffect(() => {
    function keyListener(e) {
      const listener = keyListenersMap.get(e.keyCode);
      return listener && listener(e);
    }
    document.addEventListener('keydown', keyListener);
    modalRef.current.focus()
    containerRef.current.addEventListener('click', e => clickToClose(e));

    return () => document.removeEventListener('keydown', keyListener);
  }, []);

  return (
    <Container ref={containerRef} role="dialog" aria-modal="true">
      <AnimationWrapper {...animationProps}>
        <Modal
          ref={modalRef}
          tabIndex="0" // this needs to have a tabIndex so that it can listen for the ESC key
        >
          <CloseButton onClick={() => dismissAnimationWrapper()}>
            <Cross />
          </CloseButton>
          {children}
        </Modal>
      </AnimationWrapper>
    </Container>
  );
};

SimpleModal.propTypes = {
  children: PropTypes.node.isRequired,
  closeAction: PropTypes.func.isRequired,
};

export default SimpleModal;
