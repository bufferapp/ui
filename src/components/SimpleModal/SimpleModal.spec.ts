/* eslint-disable react/jsx-filename-extension */
import snap from 'jest-auto-snapshots'
import 'jest-styled-components'

import SimpleModal from './SimpleModal'

snap(SimpleModal, './SimpleModal.jsx')
