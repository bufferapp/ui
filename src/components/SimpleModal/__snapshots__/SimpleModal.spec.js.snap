// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`jest-auto-snapshots > SimpleModal Matches snapshot when passed only required props 1`] = `
<ForwardRef(styled.div)
  aria-modal="true"
  role="dialog"
>
  <AnimationWrapper
    align="center"
    dismissing={false}
    duration={400}
    easing="cubic-bezier(0.25, 1, 0.5, 1)"
    justify="center"
    onDismiss={[MockFunction]}
    stageInAnimation={
      e {
        "id": "sc-keyframes-bQtjaF",
        "inject": [Function],
        "name": "bQtjaF",
        "rules": "
  0% {
    transform: scale(.5);
    opacity: 0;
  }

  100% {
    transform: scale(1);
    opacity: 1;
  }
",
        "toString": [Function],
      }
    }
    stageOutAnimation={
      e {
        "id": "sc-keyframes-hsHpqF",
        "inject": [Function],
        "name": "hsHpqF",
        "rules": "
  0% {
    transform: scale(1);
    opacity: 1;
  }

  100% {
    transform: scale(.5);
    opacity: 0;
  }
",
        "toString": [Function],
      }
    }
  >
    <ForwardRef(styled.div)
      tabIndex="0"
    >
      <ForwardRef(styled.button)
        onClick={[Function]}
      >
        <CrossIcon />
      </ForwardRef(styled.button)>
      <NodeFixture />
    </ForwardRef(styled.div)>
  </AnimationWrapper>
</ForwardRef(styled.div)>
`;
